text/0000-roadmap-2018.md

```diff
@@ -0,0 +1,411 @@
+- Apr — July: **Buckling down**.
+- Aug — Nov: **Fun**!
+  - August 2: **Epoch in beta**
+  - September 13: **Epoch released**
+- Dec — EOY: **Reflection**.
+
+**Design and early work**. Coming off of the impl period, for the start of this year we’ll take some time to process RFCs, to plan, and to spike any late-breaking design ideas that could possibly ship in the Rust 2018 epoch release. We will, of course, continue to push on implementation and stabilization work at the same time. By the end of this period, we should have a very clear picture of what we will ship in the epoch, and what needs to happen to do that.
+
+This period culminates with the Rust Team All Hands in Berlin at the end of March.
+
+**Buckling down**. Structured much like the impl period, we will be laser-focused on making the epoch release as good as it can be. It will be a community-wide effort. Little to no design or RFC work will happen during this time. RustFest Paris and a Mozilla all-hands will happen during this time, and we will likely try to host “impl days” similar to last year.
+
+This period culminates with cutting the beta epoch release at the beginning of August, with RustConf following shortly after.
+
+**Fun**! Having pushed hard up to this point, for the remainder of the year we will relax our tight focus and allow for more exploration, as well as work on fun features like const generics, GATs, and specialization. Rust Belt Rust will happen during this period.
+
+**Reflection**. We’ll leave the month at the end of the year for retrospectives and 2019 planning, with the hope that we can start 2019 with a roadmap already in place.
+
+# Rationale, drawbacks and alternatives
+
+Here’s a high level rationale for the two major themes this year: shipping and community.
+
+## Shipping
+
+The data and discussion quickly coalesced around one of the basic contours for this year: a sense that a lot of important improvements are in the pipeline, and **what’s needed this year is a strong focus on** ***shipping***. As @nikomatsakis put it:
+
+
+> [We did good work in 2017: now we have to make sure the world knows it and can use it.](http://smallcultfollowing.com/babysteps/blog/2018/01/09/rust2018/) (@nikomatsakis)
+
+Or, in @nrc’s more provocative phrasing:
+
+
+> [I want 2018 to be boring.](https://www.ncameron.org/blog/rust-2018/) (@nrc)
+
+This sentiment is the foundation for this year’s technical work, and there’s a striking similarity to the spirit leading up to Rust 1.0. While shipping what’s already in our pipeline may be “boring”, it’s not easy. One of the hardest challenges for us as a community is the same one we faced with 1.0, one that @skade voiced succinctly:
+
+
+> [I’d like us to stabilise imperfect things.](http://yakshav.es/rust-2018/) (@skade)
+
+Shipping requires the discipline to say “no” to exciting new ideas, and to say “yes” to delivering on ones we fear are flawed. We have to iterate toward perfection, providing useful snapshots along the way.
+
```

Comment:
In the context of libraries, this is fine since we can always make breaking changes to libraries in the ecosystem and move to new ones and *iterate* as you've described.

However, in the context of language additions, we have to be more keen on get it right the first time in my view when stabilizing because those additions will be with us for a long long time, and we shouldn't *overly* rely on epochs to fix mistakes.

If you agree, could you clarify this? =)

Reply:
Indeed! But as with all things, it's a balance. It's easy to keep anticipating future advances and never ship. [@skade's blog post](http://yakshav.es/rust-2018/) I think made some very strong points here.

(And note that the same problems arise for foundational libraries, where major version bumps are extremely disruptive.)

Reply:
For sure it is a balance; The way I read @skade's blog post, I understood it to only be about *libraries* and not T-lang related. On scale, I think shipping imperfect language features has the potential to be more disruptive however. Of course, even for language additions, you need to make a decision at some point and reach consensus after lengthy discussions. I just want to ensure that we don't leap into language additions cavalierly, which I don't think we will, since **we have the best people** ;)

Reply:
@centril Yes, it was mostly about libraries, where a second iteration is easy. When talking about language features, there should obviously be a lot of assesment whether it blocks future paths or it has a reasonable chance to ever be finished. This is very gradual, for example, in the case of a "just" API's (which block a lot of embedded things), I can still see multiple iterations happening there, _but_, obviously, you'll pay a big price of having to support them all until kindom come, even the not-so-nice ones.